src/electron/ipc.ts

```diff
@@ -0,0 +1,265 @@
+//   1. Add a `XXX_CHANNEL` constant below
+//   2. Copy the `XXX_CHANNEL` to either `OUTLINE_ONE_WAY_IPC_CHANNELS` or
+//      `OUTLINE_TWO_WAY_IPC_CHANNELS`
+//   3. Define the parameters and return type in either `OutlineIpcOneWayService`
+//      or `OutlineIpcTwoWayService`
+//   4. That's it, implement the handler, it should already be strongly typed.
+
+export const QUIT_APP_CHANNEL = 'quit-app';
+export const APP_UPDATE_DOWNLOADED_CHANNEL = 'update-downloaded';
+export const PUSH_CLIPBOARD_CHANNEL = 'push-clipboard';
+export const INSTALL_SERVICES_CHANNEL = 'install-outline-services';
+export const ADD_SERVER_CHANNEL = 'add-server';
+export const CHECK_REACHABLE_CHANNEL = 'is-server-reachable';
+export const START_VPN_CHANNEL = 'start-proxying';
+export const STOP_VPN_CHANNEL = 'stop-proxying';
+export const VPN_CONNECTED_CHANNEL = 'proxy-connected';
+export const VPN_RECONNECTING_CHANNEL = 'proxy-reconnecting';
+export const VPN_DISCONNECTED_CHANNEL = 'proxy-disconnected';
+export const REQUEST_LOCALIZATION_CHANNEL = 'localization-request';
+export const LOCALIZATION_RESPONSE_CHANNEL = 'localization-response';
+
+/**
+ * This array defines all one-way IPC channels used by `.send` and `.on`.
+ */
+export const OUTLINE_ONE_WAY_IPC_CHANNELS = <const>[
+  QUIT_APP_CHANNEL,
+  APP_UPDATE_DOWNLOADED_CHANNEL,
+  PUSH_CLIPBOARD_CHANNEL,
+  ADD_SERVER_CHANNEL,
+  VPN_CONNECTED_CHANNEL,
+  VPN_RECONNECTING_CHANNEL,
+  VPN_DISCONNECTED_CHANNEL,
+  REQUEST_LOCALIZATION_CHANNEL,
+  LOCALIZATION_RESPONSE_CHANNEL,
+];
+
+/**
+ * This array defines all two-way IPC channels used by `.invoke` and `.handle`.
+ */
+export const OUTLINE_TWO_WAY_IPC_CHANNELS = <const>[
+  INSTALL_SERVICES_CHANNEL,
+  CHECK_REACHABLE_CHANNEL,
+  START_VPN_CHANNEL,
+  STOP_VPN_CHANNEL,
+];
+
+/**
+ * Define the contract (request and response type) of all one-way IPC channels.
+ * This interface should only be used for type definitions, please **do not**
+ * try to implement it.
+ *
+ * Please make sure the return type is always `void`.
+ */
+export interface OutlineIpcOneWayService {
+  [QUIT_APP_CHANNEL]: () => void;
+  [APP_UPDATE_DOWNLOADED_CHANNEL]: () => void;
+  [PUSH_CLIPBOARD_CHANNEL]: () => void;
+  [ADD_SERVER_CHANNEL]: (url: string) => void;
+  [VPN_CONNECTED_CHANNEL]: (id: string) => void;
+  [VPN_RECONNECTING_CHANNEL]: (id: string) => void;
+  [VPN_DISCONNECTED_CHANNEL]: (id: string) => void;
+  [REQUEST_LOCALIZATION_CHANNEL]: (localizationKeys: string[]) => void;
+  [LOCALIZATION_RESPONSE_CHANNEL]: (result?: {[key: string]: string}) => void;
+}
+
+/**
+ * Define the contract (request and response type) of all two-way IPC channels.
+ * This interface should only be used for type definitions, please **do not**
+ * try to implement it.
+ *
+ * Please make sure the return type is always a `Promise` of a two-element
+ * tuple `[ErrorCode, any]` where the second `any` type can be replaced by any
```

Comment:
Maybe not necessary? We can return `object` instead of a tuple and wrap the `ErrorCode` marshaling logic in `invoke`/`handle`.